core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp

```diff
@@ -359,102 +359,103 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
   using ValueJoin   = Kokkos::Impl::FunctorValueJoin<FunctorType, WorkTag>;
   using ValueOps    = Kokkos::Impl::FunctorValueOps<FunctorType, WorkTag>;
 
+  using value_type     = typename ValueTraits::value_type;
   using pointer_type   = typename ValueTraits::pointer_type;
   using reference_type = typename ValueTraits::reference_type;
 
   const FunctorType m_functor;
   const Policy m_policy;
-  /*
-    template< class TagType >
-    inline static
-    typename std::enable_if< std::is_same< TagType , void >::value >::type
-    exec_range( const FunctorType & functor
-              , const Member ibeg , const Member iend
-              , reference_type update , const bool final )
+
+ public:
+  inline void impl_execute(
+      Kokkos::View<value_type**, Kokkos::LayoutRight,
+                   Kokkos::Experimental::OpenMPTargetSpace>
+          element_values,
+      Kokkos::View<value_type*, Kokkos::Experimental::OpenMPTargetSpace>
+          chunk_values,
+      Kokkos::View<int64_t, Kokkos::Experimental::OpenMPTargetSpace> count)
+      const {
+    const int64_t N        = m_policy.end() - m_policy.begin();
+    const int chunk_size   = 128;
+    const int64_t n_chunks = (N + chunk_size - 1) / chunk_size;
+    int nteams             = n_chunks > 512 ? 512 : n_chunks;
+    int team_size          = 128;
+
+    FunctorType a_functor(m_functor);
+#pragma omp target teams distribute map(to                             \
+                                        : a_functor) num_teams(nteams) \
+    thread_limit(team_size)
+    for (int64_t team_id = 0; team_id < n_chunks; team_id++) {
+#pragma omp parallel num_threads(team_size)
       {
-        #ifdef KOKKOS_OPT_RANGE_AGGRESSIVE_VECTORIZATION
-        #ifdef KOKKOS_ENABLE_PRAGMA_IVDEP
-        #pragma ivdep
-        #endif
-        #endif
-        for ( Member iwork = ibeg ; iwork < iend ; ++iwork ) {
-          functor( iwork , update , final );
+        int64_t local_offset = team_id * chunk_size;
+
+#pragma omp for
+        for (int i = 0; i < chunk_size; i++) {
+          int64_t idx = local_offset + i;
+          if (idx < N) a_functor(idx, element_values(team_id, i), false);
+        }
+#pragma omp barrier
+        if (omp_get_thread_num() == 0) {
+          value_type sum = 0;
+          for (int i = 0; i < chunk_size; i++) {
+            sum += element_values(team_id, i);
+            element_values(team_id, i) = sum;
```

Comment:
`ValueOps::copy`?

Reply:
we don't support array type scans so I think its not needed.